### ISSUE TITLE
Worker stops after queuing several jobs in a row

### ISSUE THREAD

**Issue by mac3851** [msolli/proletarian#5]
I've tried to enqueue several jobs in a row like:
```
(doseq [{:keys [email]} users]
    (when-not (empty? email)
      (worker/enqueue! worker :send-email {:email email})))
```
This resulted in:
```
:proletarian.worker/polling-for-jobs {:worker-thread-id 1, :proletarian.worker/queue-worker-id proletarian[:proletarian/default]}
:proletarian.worker/job-worker-error {:throwable #error {
 :cause HikariPool-1 - Connection is not available, request timed out after 30000ms.
 :via
 [{:type java.sql.SQLTransientConnectionException
   :message HikariPool-1 - Connection is not available, request timed out after 30000ms.
   :at [com.zaxxer.hikari.pool.HikariPool createTimeoutException HikariPool.java 676]}]
 :trace
 [[com.zaxxer.hikari.pool.HikariPool createTimeoutException HikariPool.java 676]
  [com.zaxxer.hikari.pool.HikariPool getConnection HikariPool.java 190]
  [com.zaxxer.hikari.pool.HikariPool getConnection HikariPool.java 155]
  [com.zaxxer.hikari.HikariDataSource getConnection HikariDataSource.java 100]
  [net.ttddyy.dsproxy.support.ProxyDataSource getConnection ProxyDataSource.java 63]
  [proletarian.db$with_connection invokeStatic db.clj 128]
  [proletarian.db$with_connection invoke db.clj 127]
  [proletarian.db$with_tx invokeStatic db.clj 132]
  [proletarian.db$with_tx invoke db.clj 131]
  [proletarian.worker$process_next_job_BANG_ invokeStatic worker.clj 44]
  [proletarian.worker$process_next_job_BANG_ invoke worker.clj 11]
  [proletarian.worker$process_next_jobs_BANG_ invokeStatic worker.clj 74]
  [proletarian.worker$process_next_jobs_BANG_ invoke worker.clj 66]
  [proletarian.worker$create_queue_worker$reify__45271$work_BANG___45277 invoke worker.clj 190]
  [clojure.core$partial$fn__5839 invoke core.clj 2623]
  [clojure.lang.AFn run AFn.java 22]
  [java.util.concurrent.Executors$RunnableAdapter call Executors.java 515]
  [java.util.concurrent.FutureTask runAndReset FutureTask.java 305]
  [java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask run ScheduledThreadPoolExecutor.java 305]
  [java.util.concurrent.ThreadPoolExecutor runWorker ThreadPoolExecutor.java 1128]
  [java.util.concurrent.ThreadPoolExecutor$Worker run ThreadPoolExecutor.java 628]
  [java.lang.Thread run Thread.java 829]]}, :proletarian.worker/queue-worker-id proletarian[:proletarian/default]}
:proletarian.executor/shutting-down {:proletarian.worker/queue-worker-id proletarian[:proletarian/default]}
:proletarian.executor/completed-shutdown {:proletarian.worker/queue-worker-id proletarian[:proletarian/default]}
```
On the other side, I've tried to manually insert several jobs directly to `proletarian.job` table like:
```
INSERT INTO proletarian.job (job_id,queue,job_type,payload,attempts,enqueued_at,process_at) VALUES 
('a7f4a6f3-1aaf-4fbd-aca9-731a71fa8a64',':proletarian/default',':send-email','["^ ","~:email","email1@email.com"]',0,'2021-10-29 16:27:37.756','2021-10-29 16:27:37.756')
,('47d2ed6e-a7d3-4d9c-97ca-ac6c2ad8c541',':proletarian/default',':send-email','["^ ","~:email","email2@email.com"]',0,'2021-10-29 16:27:37.780','2021-10-29 16:27:37.780')
,('7763a48d-92fb-4164-ae79-e4b4e688b1ef',':proletarian/default',':send-email','["^ ","~:email","email3@email.com"]',0,'2021-10-29 16:27:37.796','2021-10-29 16:27:37.796')
,('51c87997-a9b2-4a0c-b852-3a4fbf0fb20a',':proletarian/default',':send-email','["^ ","~:email","email4@email.com"]',0,'2021-10-29 16:27:37.821','2021-10-29 16:27:37.821')
,('29e63f4c-7d3d-4f03-afc3-2762e108b8a7',':proletarian/default',':send-email','["^ ","~:email","email5@email.com"]',0,'2021-10-29 16:27:37.837','2021-10-29 16:27:37.837')
,('3d3747db-4c5a-4a9b-ba24-b3dced35056c',':proletarian/default',':send-email','["^ ","~:email","email6@email.com"]',0,'2021-10-29 16:27:37.845','2021-10-29 16:27:37.845')
,('daa631c9-2cf4-496b-9876-c8cf7f968be7',':proletarian/default',':send-email','["^ ","~:email","email7@email.com"]',0,'2021-10-29 16:27:37.860','2021-10-29 16:27:37.860')
,('b862c0d0-4fa0-48bc-a024-46563401a84c',':proletarian/default',':send-email','["^ ","~:email","email8@email.com"]',0,'2021-10-29 16:27:37.876','2021-10-29 16:27:37.876')
,('2b7dc9fb-500b-4cc3-877a-712f8aacdb70',':proletarian/default',':send-email','["^ ","~:email","email9@email.com"]',0,'2021-10-29 16:27:37.894','2021-10-29 16:27:37.894')
,('10b6a4e7-1d22-4adb-8781-63686d3e3c90',':proletarian/default',':send-email','["^ ","~:email","email10@email.com"]',0,'2021-10-29 16:27:37.903','2021-10-29 16:27:37.903');
```
This worked like a charm and all jobs were successfully processed.

Are you able to take a look into it or implement something like `enqueue-many!`?

**Comment by msolli** [comment]
I noticed you're using HikariCP for connection pooling. I can recreate the error you have by setting the `:maximum-pool-size` for HikariCP to the same value as `:proletarian/worker-threads`. Then, if you enqueue jobs using the same datasource, there might not be a connection available for the workers when they poll.

I'm going to look into having the workers retry instead of giving up when they encounter a `java.sql.SQLTransientException`.

In the meantime I'd encourage you to use a separate datasource for the workers. With HikariCP you'd set up one with the same value for `:maximum-pool-size` as `:proletarian/worker-threads`, making sure no-one else uses this connection pool. Then you'd always have a connection available when polling. What are the values you're using that is causing problems?

**Comment by msolli** [comment]
There's a new commit on main for you to try: https://github.com/msolli/proletarian/commit/0bf9dff6c2c99a3f17a914ce582631133ce1b93f

Use these coordinates in `deps.edn`:
```clojure
msolli/proletarian {:git/url "https://github.com/msolli/proletarian"
                    :sha     "0bf9dff6c2c99a3f17a914ce582631133ce1b93f"}
```

This will not stop the Queue Worker when encountering a SQLTransientException, just log it. The polling will continue.

This does not solve your underlying problem, though, which is that there aren't enough connections in the connection pool for your workers and your enqueuing code (see the previous comment).

**Comment by msolli** [comment]
The fix is now available in release [v1.0.54-alpha](https://github.com/msolli/proletarian/releases/tag/v1.0.54-alpha).
